lib/SPIRV/SPIRVWriter.cpp

```diff
@@ -626,17 +626,74 @@ SPIRVType *LLVMToSPIRVBase::transSPIRVJointMatrixINTELType(
   else
     llvm_unreachable("Unexpected type for matrix!");
 
-  auto ParseInteger = [this](StringRef Postfix) -> ConstantInt * {
+  auto ParseInteger = [](StringRef Postfix) -> uint32_t {
     unsigned long long N = 0;
     consumeUnsignedInteger(Postfix, 10, N);
-    return getUInt32(M, N);
+    return N;
   };
-  std::vector<SPIRVValue *> Args;
+  std::vector<uint32_t> ProcessedPostfixes;
   for (size_t I = 1; I != Postfixes.size(); ++I)
-    Args.emplace_back(transConstant(ParseInteger(Postfixes[I])));
+    ProcessedPostfixes.emplace_back(ParseInteger(Postfixes[I]));
+  // Overwrite the 'Layout' argument depending on 'Use' if we have it.
+  if (ProcessedPostfixes.size() == 5) {
+    if (ProcessedPostfixes[4] == internal::InternalJointMatrixUse::MatrixB)
+      ProcessedPostfixes[2] = internal::InternalJointMatrixLayout::PackedB;
+    else
+      ProcessedPostfixes[2] = internal::InternalJointMatrixLayout::RowMajor;
+  }
+  std::vector<SPIRVValue *> Args;
+  for (auto Val : ProcessedPostfixes)
+    Args.emplace_back(transConstant(getUInt32(M, Val)));
   return BM->addJointMatrixINTELType(transType(ElemTy), Args);
 }
 
+// Helper for translation of OpJointMatrixLoadINTEL and OpJointMatrixStoreINTEL
+// instructions.
+// For 'Layout' we have the following enum in SYCL spec:
+// { row_major, col_major, packed, unused }
+// while the SPIR-V spec states, that 'Layout' can be:
+// { RowMajor, ColMajor, PackedA, PackedB }
+// So we need to transform packed to PackedA or PackedB depending on
+// 'Use' parameter of the JointMatrix type.
+template<uint32_t LayoutPos = 0>
+SPIRVValue *LLVMToSPIRVBase::transSPIRVJointMatrixINTELMemoryLayout(
+    CallInst *CI, SPIRVBasicBlock *BB, SPIRVTypeJointMatrixINTEL *MatTy) {
+  auto *Use = MatTy->getUse();
+  if (!Use)
```

Comment:
if we don't have use and layout is packed, layout is still "packed" after spv-writer?
@MrSidims

Reply:
Yes. I consider, that for previous code cases we still have PackedA and PackedB as the layout. It will remain the same.